drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,1058 @@
+	desc->axi_addr = axi_addr;
+	desc->length = size;
+#ifdef CONFIG_DMA_IPROC_PAX_TOGGLE_MODE
+	desc->toggle = toggle;
+#elif CONFIG_DMA_IPROC_PAX_DOORBELL_MODE
+	desc->toggle = 0;
+#endif
+
+	if (direction == CARD_TO_HOST) {
+		desc->type = is_mega ?
+			     PAX_DMA_TYPE_MEGA_SRC_DESC : PAX_DMA_TYPE_SRC_DESC;
+	} else {
+		desc->type = is_mega ?
+			     PAX_DMA_TYPE_MEGA_DST_DESC : PAX_DMA_TYPE_DST_DESC;
+	}
+}
+
+#ifdef CONFIG_DMA_IPROC_PAX_TOGGLE_MODE
+static void INIT_TOGGLE(void *desc, uint32_t toggle)
+{
+	struct rm_header *r = (struct rm_header *)desc;
+
+	r->toggle = toggle;
+}
+#endif
+
+static inline void *__CURR_INC(struct dma_iproc_pax_rm_ring_data *ring)
+{
+	struct next_ptr_desc *nxt;
+	uintptr_t curr;
+
+	curr = (uintptr_t)ring->curr.write_ptr;
+	/* if hit next table ptr, skip to next location, flip toggle */
+	nxt = (struct next_ptr_desc *)curr;
+	if (nxt->type == PAX_DMA_TYPE_NEXT_PTR) {
+		LOG_DBG("hit next_ptr@0x%lx %d, next_table@0x%lx\n",
+			curr, nxt->toggle, (uintptr_t)nxt->addr);
+		uintptr_t last = (uintptr_t)ring->bd +
+			     PAX_DMA_RM_DESC_RING_SIZE * PAX_DMA_NUM_BD_BUFFS;
+		nxt->toggle = ring->curr.toggle;
+		ring->curr.toggle = (ring->curr.toggle == 0) ? 1 : 0;
+		/* move to next addr, wrap around if hits end */
+		curr += PAX_DMA_RM_DESC_BDWIDTH;
+		if (curr == last) {
+			curr = (uintptr_t)ring->bd;
+			LOG_DBG("hit end of desc:0x%lx, wrap to 0x%lx\n",
+				last, curr);
+		}
+		ring->descs_inflight++;
+	}
+
+	ring->curr.write_ptr = (void *)(curr + PAX_DMA_RM_DESC_BDWIDTH);
+	ring->descs_inflight++;
+
+	return (void *)curr;
+}
+
+static void NEXT_TABLE_DESC(void *desc, void *next_ptr, uint32_t toggle)
+{
+	struct next_ptr_desc *nxt = (struct next_ptr_desc *)desc;
+
+	nxt->addr = (uintptr_t)next_ptr;
+	nxt->type = PAX_DMA_TYPE_NEXT_PTR;
+	nxt->toggle = toggle;
+}
+
+static inline void *__CURR(struct dma_iproc_pax_rm_ring_data *ring)
+{
+	return ring->curr.write_ptr;
+}
+
+static void prepare_ring(struct dma_iproc_pax_rm_ring_data *ring)
+{
+	uintptr_t curr, next, last;
+	int buff_count = PAX_DMA_NUM_BD_BUFFS;
+#ifdef CONFIG_DMA_IPROC_PAX_TOGGLE_MODE
+	uint32_t toggle;
+#endif
+
+	/* zero out descriptor area */
+	memset(ring->bd, 0x0, PAX_DMA_RM_DESC_RING_SIZE * PAX_DMA_NUM_BD_BUFFS);
+	memset(ring->cmpl, 0x0, PAX_DMA_RM_CMPL_RING_SIZE);
+
+	/* start with first buffer, valid toggle is 0x1 */
+#ifdef CONFIG_DMA_IPROC_PAX_TOGGLE_MODE
+	toggle = 0x1;
+#endif
+	curr = (uintptr_t)ring->bd;
+	next = curr + PAX_DMA_RM_DESC_RING_SIZE;
+	last = curr + PAX_DMA_RM_DESC_RING_SIZE * PAX_DMA_NUM_BD_BUFFS;
+	do {
+#ifdef CONFIG_DMA_IPROC_PAX_TOGGLE_MODE
+		INIT_TOGGLE((void *)curr, toggle);
+		/* Place next_table desc as last BD entry on each buffer */
+		NEXT_TABLE_DESC(PAX_DMA_NEXT_TBL_ADDR((void *)curr),
+				(void *)next, toggle);
+#elif CONFIG_DMA_IPROC_PAX_DOORBELL_MODE
+		/* Place next_table desc as last BD entry on each buffer */
+		NEXT_TABLE_DESC(PAX_DMA_NEXT_TBL_ADDR((void *)curr),
+				(void *)next, 0);
```

Comment:
Just for information purpose, I am using the door bell mode, but I am also setting toggle bit in my driver, is there is negative effects?

Reply:
There's no side effect, RM will ignore the toggle bit.